runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -2610,14 +2610,26 @@ class INTERPRETER_CLASS
 		_sp += (slotCount - 1);
 	}
 
-	/* java.lang.Thread: public static native Thread _currentThread(); */
+	/* java.lang.Thread: public static native Thread currentThread(); */
 	VMINLINE VM_BytecodeAction
 	inlThreadCurrentThread(REGISTER_ARGS_LIST)
 	{
 		returnObjectFromINL(REGISTER_ARGS, _currentThread->threadObject, 0);
 		return EXECUTE_BYTECODE;
 	}
 
+#if JAVA_SPEC_VERSION >= 19
+	/* java.lang.Thread: native void setCurrentThread(Thread thread); */
```

Comment:
Should this not be static?

Reply:
> Are we sure it's only called with the current thread?

The API doc states _Sets the Thread object to be returned by Thread.currentThread()._ [1] https://github.com/eclipse-openj9/openj9/pull/15428#discussion_r907615278
My interpretation is that it means to change current thread.
It appears `static` makes sense, there weren't much hints at the initial commit - https://github.com/openjdk/loom/commit/69a994c698b58fffb695890ee8ab02f2e80d9234

Reply:
Static ensures that it can only possibly apply to the current thread without passing in a pointless parameter that can only have one valid value.

I'd suggest asking the Loom group - this feel like a foolish oversight. This does not affect the JNI versions of the natives, but it will affect the INL as statics do not pass the class to INLs.

Reply:
I suppose it depends on the scheduler implementation - do carrier threads mount their own vthreads or is it done by another thread (in which case, the non-static version makes sense, though it would be difficult to implement reliably).

Reply:
Roughly as the non-static scenario you thought, when a `Callable` task is to be executed on carrier thread,
```
currentThread = Thread.currentThread() 
if currentThread is a virtual thread, 
    Thread carrierThread = Thread.currentCarrierThread();
    carrierThread.setCurrentThread(carrier);
    run the task
    carrierThread.setCurrentThread(thread);
```
So the `setCurrentThread` is invoked via a carrierThread instance which is going to set the thread returning by Thread.currentThread().

Reply:
In that case, the existing native implementation is incorrect, so the new implementations are as well. You will need to set the threadObject on the receiver, not the current thread.

Reply:
Fixed, please have another look.